onmt/io/AudioDataset.py

```diff
@@ -283,3 +290,122 @@ def get_num_features(corpus_file, side):
                 _, _, num_feats = AudioDataset.extract_text_features(f_line)
 
         return num_feats
+
+
+class ShardedAudioCorpusIterator(object):
+    """
+    This is the iterator for audio corpus, used for sharding large audio
+    corpus into small shards, to avoid hogging memory.
+
+    Inside this iterator, it automatically divides the audio files into
+    shards of size `shard_size`. Then, for each shard, it processes
+    into (example_dict, n_features) tuples when iterates.
+    """
+    def __init__(self, src_dir, corpus_path, truncate, side, shard_size,
+                 sample_rate, window_size, window_stride,
+                 window, normalize_audio=True, assoc_iter=None):
+        """
+        Args:
+            corpus_path: the corpus file path.
+            truncate: .
+            side: "src" or "tgt".
+            shard_size: the shard size, 0 means not sharding the file.
+            assoc_iter: if not None, it is the associate iterator that
+                        this iterator should align its step with.
+        """
+        try:
+            # The codecs module seems to have bugs with seek()/tell(),
+            # so we use io.open().
+            self.corpus = io.open(corpus_path, "r", encoding="utf-8")
+        except IOError:
+            sys.stderr.write("Failed to open corpus file: %s" % corpus_path)
+            sys.exit(1)
+
+        self.side = side
+        self.src_dir = src_dir
+        self.shard_size = shard_size
+        self.sample_rate = sample_rate
+        self.truncate = truncate
+        self.window_size = window_size
+        self.window_stride = window_stride
+        self.window = window
+        self.normalize_audio = normalize_audio
+        self.assoc_iter = assoc_iter
+        self.last_pos = 0
+        self.last_line_index = -1
+        self.line_index = -1
+        self.eof = False
+
+    def __iter__(self):
+        """
+        Iterator of (example_dict, nfeats).
+        On each call, it iterates over as many (example_dict, nfeats) tuples
+        until this shard's size equals to or approximates `self.shard_size`.
+        """
+        iteration_index = -1
+        if self.assoc_iter is not None:
+            # We have associate iterator, just yields tuples
+            # util we run parallel with it.
+            while self.line_index < self.assoc_iter.line_index:
+                line = self.corpus.readline()
+                if line == '':
+                    raise AssertionError(
+                        "Two corpuses must have same number of lines!")
+
+                self.line_index += 1
+                iteration_index += 1
+                yield self._example_dict_iter(line, iteration_index)
+
+            if self.assoc_iter.eof:
+                self.eof = True
+                self.corpus.close()
+        else:
```

Comment:
The original associate iterator is to synchronize the shards between source text file and target  text file. Concretely, we use `max_shard_size` to determine how many lines a shard of source text file contains, then make the target side synchronize with it, using the `assoc_iter`.

IIUC,  for Audio corpus, the source side is a specific audio file, thus the target side shall be the corresponding target text lines, so I think the `assoc_iter` logic should change to reflect this fact?

Also stale comment to update.